src/client/driver/command-executors/execute-action-command.js

```diff
@@ -2,110 +2,126 @@ import hammerhead from '../deps/hammerhead';
-        .catch(err => {
-            progressPanel.close(false);
-            throw err;
+            return ensureElementVisible(element, checkVisibilityTimeout, IsInvisibleErrorCtor);
         });
 }
 
-function runAutomation (element, command) {
-    var automation = null;
+function createAutomation (elements, command) {
+    /* eslint-disable indent*/
+    // TODO: eslint raises an 'incorrect indent' error here. We use
+    // the old eslint version (v1.x.x). We should migrate to v2.x.x
+    switch (command.type) {
+        case COMMAND_TYPE.click :
+            return new ClickAutomation(elements[0], command.options);
 
-    if (command.type === COMMAND_TYPE.click)
-        automation = new ClickAutomation(element, command.options);
+        case COMMAND_TYPE.rightClick :
+            return new RClickAutomation(elements[0], command.options);
 
-    if (command.type === COMMAND_TYPE.rightClick)
-        automation = new RClickAutomation(element, command.options);
+        case COMMAND_TYPE.doubleClick :
+            return new DblClickAutomation(elements[0], command.options);
 
-    if (command.type === COMMAND_TYPE.doubleClick)
-        automation = new DblClickAutomation(element, command.options);
+        case COMMAND_TYPE.drag :
+            return new DragToOffsetAutomation(elements[0], command.dragOffsetX, command.dragOffsetY, command.options);
 
-    if (command.type === COMMAND_TYPE.hover)
-        automation = new HoverAutomation(element, command.options);
+        case COMMAND_TYPE.dragToElement :
+            return new DragToElementAutomation(elements[0], elements[1], command.options);
 
-    return automation
-        .run()
-        .then(() => {
-            return { failed: false };
-        });
+        case COMMAND_TYPE.hover :
+            return new HoverAutomation(elements[0], command.options);
+    }
+    /* eslint-enable indent*/
 }
 
+
 export default function executeActionCommand (command) {
     var resolveStartPromise = null;
     var startPromise        = new Promise(resolve => resolveStartPromise = resolve);
 
     var completePromise = new Promise(resolve => {
         var xhrBarrier    = null;
         var commandResult = null;
-        var selector      = () => window.eval(command.selector);
 
-        ensureElement(selector)
-            .then(element => {
+        var progressPanel = new ProgressPanel();
+
+        progressPanel.show(PROGRESS_PANEL_TEXT, CHECK_ELEMENT_TIMEOUT);
+
+        var ensureElementPromises = [ensureElement(command.selector, ActionElementNotFoundError, ActionElementIsInvisibleError)];
+
+        if (command.type === COMMAND_TYPE.dragToElement)
+            ensureElementPromises.push(ensureElement(command.destinationSelector, DragDestinationNotFoundError, DragDestinationIsInvisibleError));
+
+        Promise.all(ensureElementPromises)
+            .catch(err => {
+                progressPanel.close(false);
+
+                throw err;
+            })
+            .then(elements => {
```

Comment:
Can we refactor whole ensure element part (including progress panel) to the separate method?